staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,595 @@
+		if req == nil {
+			metrics.AddReject(qs.name, "queue-full")
+			return true, false, false, func() {}
+		}
+		// ========================================================================
+
+		// ------------------------------------------------------------------------
+		// Step 2:
+		// 1) The next step is to invoke the method that dequeues as much as possible.
+
+		// This method runs a loop, as long as there
+		// are non-empty queues and the number currently executing is less than the
+		// assured concurrency value.  The body of the loop uses the fair queuing
+		// technique to pick a queue, dequeue the request at the head of that
+		// queue, increment the count of the number executing, and send true to
+		// the request's channel.
+		qs.dequeueWithChannelAsMuchAsPossible()
+		// ------------------------------------------------------------------------
+		return false, false, false, func() {}
+	}()
+	if shouldReturn {
+		return tryAnother, execute, afterExecution
+	}
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.name, "time-out")
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request cancelled\n")
+		func() {
+			qs.lockAndSyncTime()
+			defer qs.lock.Unlock()
+
+			// TODO(aaron-prindle) add metrics to these two cases
+			if req.Enqueued {
+				// remove the request from the queue as it has timed out
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						// remove the request
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+					qs.getRequestsExecutingLocked() == 0 {
+					qs.counter.Add(1)
+					go func(eh fq.EmptyHandler) {
+						eh.HandleEmpty()
+						qs.counter.Add(-1)
+					}(qs.emptyHandler)
+				}
+			} else {
+				// At this point we know that req was in its queue earlier and another
+				// goroutine has removed req from its queue and called qs.counter.Add(1)
+				// in anticipation of unblocking this goroutine through the other arm of this
+				// select. In this case we need to decrement the counter because this goroutine
+				// was actually unblocked through a different code path.
+				qs.counter.Add(-1)
```

Comment:
This might be a character flaw, but I'm just going to admit, I'm extremely suspicious of code that has counter increments/decrements scattered about (same with locks / unlocks). It is very hard to manually audit, and I don't know how you can write a test that will convince me that it's 100% correct.

Reply:
I am not happy about this one either.  It is here because we are doing a join: two separate pieces of code do an increment and an action intended to unblock this goroutine, but obviously this goroutine actually gets unblocked only once.

Reply:
See suggested improvement in standalone comments

Reply:
Suppose we made two methods on the queueSet:

```
func (qs *queueSet) preUnblockThread() { qs.counter.Add(1) }

func (qs *queueSet) threadIsFinishedOrBlocked() { qs.counter.Add(-1) }
```

That would make this code self-documenting and several times easier to manually audit, I think?